README.md
# InstagramBot.js
[![License](https://img.shields.io/badge/license-GLPv3-brightgreen.svg)]()
[![powered by webdriverio](https://img.shields.io/badge/powered%20by-webdriverio-46aef7.svg)](https://github.com/webdriverio/webdriverio)
[![Version](https://img.shields.io/badge/version-v0.3%20BETA-lightgrey.svg)](https://github.com/ptkdev/instagram-bot.js/releases)
[![Slack Chat](https://img.shields.io/badge/chat%20on-Slack-orange.svg)](https://slack.ptkdev.io)
[![Paypale Donate](https://img.shields.io/badge/donate-PayPal-red.svg)](https://paypal.me/ptkdev)

[![https://instagram-bot.js.ptkdev.io](https://ptkdev.it/img/bot/ptkdev-instagram-bot.gif)](https://instagram-bot.js.ptkdev.io)

# Features
* [✓] Login
* [✓] 2FA
* [✓] Multi-Session
* [✓] Errors manager (bad pin, bad password)
* [✓] Screenshot and Verbose logger
* [✓] Like Mode Classic: bot select random hashtag from config list and like 1 random photo (of last 20), and repeat this all time | 750-900 like/day. Limit is 1000/day for ig. This is safe mode.

# Setup - Debian Server
### Install bot dependencies:
1. `sudo apt-get install default-jre build-essential xvfb libssl-dev curl wget git chromium xauth`

### Install google chrome v63
1. `sudo dpkg -i ./bin/google-chrome-stable_current_amd64.deb`
2. `sudo apt-get install -y -f`

If you need lastest version of chrome update binary files in bin folder:
1. `/bin/chromedriver` [Download](https://sites.google.com/a/chromium.org/chromedriver/)
2. `/bin/selenium-server-standalone.jar` [Download](http://www.seleniumhq.org/download/)

### Install Node and bot dependencies
1. `curl -sL https://deb.nodesource.com/setup_8.x -o nodesource_setup.sh `
2. `sudo bash nodesource_setup.sh`
3. `rm nodesource_setup.sh`
4. `sudo apt-get install nodejs`
5. `npm install`

## Run Selenium
If you have desktop environment (example: gnome) run:
1. `npm run start-selenium-desktop`

If you not have a desktop environment (example: run bot on server) run:
1. `npm run start-selenium-server`

#### Check if work, run:
1. `pm2 logs`
2. If last line is `INFO - Selenium Server is up and running` selenium work great.

BUG: if you received `Unable to access jarfile` edit `sh` files in `bin folder` and add full path of binary, example:

`java -Dwebdriver.chrome.driver="/home/your_name/instagram-bot.js/bin/chromedriver" -jar /home/your_name/instagram-bot.js/bin/selenium-server-standalone.jar`

## Run Bot
1. Copy root file `config.js.tpl` to `config.js`, fill it properly.
2. Start the bot via `npm run start` or pm2 `npm run start-pm2`

#### Pin
If you received sms or email pin edit `loginpin.txt` and insert it on first line. Wait 3 minutes...

#### Check if work:
See logs: `cat ./logs/debug.log` or png images in ./logs/screenshot

# TODO
Features:
* likemode_superlike - select random hashtag from config list and like 3 random photo of same user | 750-900 like/day.
* fdfmode_defollowall - defollow all your following (not defollow users in whitelist) | 90 defollow/hour.
* fdfmode_classic - follow user from random hashtag and defollow after 1h | 300 follow-defollow/day.
* Add total like/day in config.js 

Nice have:
* docker container

# Social Manager Tools
Available:
- [InstagramBot.js](https://github.com/social-manager-tools/instagram-bot.js)

Coming soon:
- [TwitterBot.js](https://github.com/social-manager-tools/twitter-bot.js)
- [FacebookPageBot.js](https://github.com/social-manager-tools/facebook-page-bot.js)
- [TelegramWordpressBot.js](https://github.com/social-manager-tools/telegram-wordpress-news-bot.js)

# License

GNU GENERAL PUBLIC LICENSE

Copyright (c) 2018 Patryk Rzucidło (PTKDev)

CHANGELOG.md
# v0.3.0 BETA (January 16, 2018)

* Refactor: moved to async wait
* Refactor: add pm2 and forever to npm run
* Update: chromedriver binary
* Fix: get url picture selector
* Fix: <3 selector
* Fix: screen server resolution
* Bug: 2FA not work but work 2FA-security location error flow


# v0.2.1 (January 15, 2018)

* Ops: bad run script in package.json


# v0.2.0 (January 15, 2018)

* New flow pattern
* Refactor: removed eval()
* Refactor: removed setTimeout() and added system-sleep (run: npm install)
* Refactor: folder routes moved to modules 
* Feature: switch_mode()
* Fix: 2FA bugs
* Fix: errors manager bugs
* Fix: screenshot bugs
* Fix: random dislike


# v0.1.1 (January 13, 2018)

* Fix debug (verbose)
* Updated README.md


# v0.1.0 (January 12, 2018)

* First Release.
* Login
* 2FA
* Multi-Session
* Errors manager (bad pin, bad password)
* Screenshot and Verbose logger
* Like Mode Classic: bot select random hashtag from config list and like 1 random photo (of last 20), and repeat this all time | 750-900 like/day. Limit is 1000/day for ig. This is safe mode.

bot.js
/**
 * InstagramBot.js
 * =====================
 * Instagram Bot made with love and nodejs
 *
 * @author:     Patryk Rzucidlo [@ptkdev] <info@ptkdev.it> (https://ptkdev.it)
 * @file:       bot.js
 * @version:    0.2
 *
 * @license:    Code and contributions have 'GNU General Public License v3'
 *              This program is free software: you can redistribute it and/or modify
 *              it under the terms of the GNU General Public License as published by
 *              the Free Software Foundation, either version 3 of the License, or
 *              (at your option) any later version.
 *              This program is distributed in the hope that it will be useful,
 *              but WITHOUT ANY WARRANTY; without even the implied warranty of
 *              MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *              GNU General Public License for more details.
 *              You should have received a copy of the GNU General Public License
 *              along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @link        Homepage:     https://instagram-bot.js.ptkdev.io
 *              GitHub Repo:  https://github.com/ptkdev/instagram-bot.js
 */

/**
 * Libs
 * =====================
 * Open source library
 *
 * @author:     Patryk Rzucidlo [@ptkdev] <info@ptkdev.it> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 * @version:    0.2
 * @link:       http://webdriver.io/
 * @changelog:  0.1 initial release
 *              0.2 refactor: removed useless vars
 *
 */
const webdriverio = require('webdriverio');
const path = require('path');
const fs = require('fs');
const config = require(__dirname + '/config');
const options = {
    desiredCapabilities: {
        browserName: config.selenium_browser,
        chromeOptions: {
            args: config.selenium_chrome_options
        },
        binary: config.selenium_chrome_path,
        host: config.selenium_host,
        port: config.selenium_port
    }
};

/**
 * Init
 * =====================
 * Get username, password and hashtag of bot from /config.js
 * If not exist rename config.js.tmpl to config.js and change strings
 *
 * @author:     Patryk Rzucidlo [@ptkdev] <info@ptkdev.it> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 * @version:    0.1
 * @changelog:  0.1 initial release
 *
 */
let client = webdriverio.remote(options);
var bot = client.init();

/**
 * Import libs
 * =====================
 * Flow and utilty modules
 *
 * @author:     Patryk Rzucidlo [@ptkdev] <info@ptkdev.it> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 * @version:    0.2
 * @changelog:  0.1 initial release
 *              0.2 refactor: removed eval() and added require.
 *
 */
let utils = require(__dirname + '/modules/utils.js')(bot, config);

/**
 * Start Bot (init vars)
 * =====================
 *
 * @author:     Patryk Rzucidlo [@ptkdev] <info@ptkdev.it> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 * @version:    0.1
 * @changelog:  0.1 initial release
 *
 */
let login_status = "";
let twofa_status = "";
let like_status = "";
let pin_status = "";

/**
 * Switch Mode
 * =====================
 * Switch mode from config.js
 * 
 * @author:     Patryk Rzucidlo [@ptkdev] <info@ptkdev.it> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 * @version:    0.1
 * @changelog:  0.1 initial release
 *
 */
async function switch_mode(bot, config, utils) {
    if (config.bot_mode == "likemode_classic")
        await start_likemode_classic(bot, config, utils);
}

/**
 * LikemodeClassic Flow
 * =====================
 * /modules/likemode_classic.js
 * 
 * @author:     Patryk Rzucidlo [@ptkdev] <info@ptkdev.it> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 * @version:    0.1
 * @changelog:  0.1 initial release
 *
 */
async function start_likemode_classic(bot, config, utils) {
    let likemode_classic = require(__dirname + '/modules/likemode_classic.js')(bot, config, utils);
    utils.logger("[INFO]", "likemode", "classic");
    let today = "";
    let t1, t2, sec;
    do {
        today = new Date();
        t1 = new Date(today.getFullYear(), today.getMonth(), today.getDate(), today.getHours(), today.getMinutes(), today.getSeconds());
        utils.logger("[INFO]", "like", "loading... " + new Date(today.getFullYear(), today.getMonth(), today.getDate(), today.getHours(), today.getMinutes(), today.getSeconds()));
        likemode_classic.like_open_hashtagpage();
        utils.sleep(utils.random_interval(4, 8));
        await likemode_classic.like_get_urlpic();
        utils.sleep(utils.random_interval(4, 8));
        like_status = await likemode_classic.like_click_heart();
        today = new Date();
        t2 = new Date(today.getFullYear(), today.getMonth(), today.getDate(), today.getHours(), today.getMinutes(), today.getSeconds());
        sec = Math.abs((t1.getTime() - t2.getTime())/ 1000);
        utils.logger("[INFO]", "likemode", "seconds of loop "+sec+"... for miss ban bot wait "+(90 - sec)+"-"+(90 - sec + 15));
        if(sec < 90)
        utils.sleep(utils.random_interval(90 - sec, (90 - sec + 15)));
    } while (true);
}

/**
 * 2FA Flow (check if work)
 * =====================
 * /modules/2FA.js
 * 
 * @author:     Patryk Rzucidlo [@ptkdev] <info@ptkdev.it> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 * @version:    0.1
 * @changelog:  0.1 initial release
 *
 */
async function start_twofa_check() {
    utils.logger("[INFO]", "twofa", "instagram request pin?");
    try {
        let attr = await bot.getAttribute('#choice_1', 'value');
        if (config.debug == true)
            utils.logger("[DEBUG]", "twofa", "attr = " + attr);
        utils.logger("[INFO]", "twofa", "yes, instagram require security pin... You can not pass!1!111! (cit.)");
        utils.screenshot(bot, "twofa", "check_pin_request");
        pin_status = 1;
    } catch (err) {
        if (config.debug == true)
            utils.logger("[DEBUG]", "twofa", err);
        utils.logger("[INFO]", "twofa", "no, bot is at work (started)... Wait...");
        utils.logger("[INFO]", "twofa", "starting current mode");
        utils.screenshot(bot, "twofa", "check_nopin");
        pin_status = 0;
    }
}

/**
 * 2FA Flow
 * =====================
 * /modules/2FA.js
 * 
 * @author:     Patryk Rzucidlo [@ptkdev] <info@ptkdev.it> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 * @version:    0.1
 * @changelog:  0.1 initial release
 *
 */
async function start_twofa() {
    let twofa = require(__dirname + '/modules/2FA.js')(bot, config, utils);
    utils.logger("[INFO]", "twofa", "loading...");
    twofa.sendpin();
    utils.sleep(utils.random_interval(170, 180));
    twofa.readpin();
    utils.sleep(utils.random_interval(4, 8));
    twofa.submit();
    utils.sleep(utils.random_interval(4, 8));
    twofa_status = await twofa.submitverify();
    utils.sleep(utils.random_interval(4, 8));
    bot = twofa.get_bot();
}

/**
 * Login Flow
 * =====================
 * /modules/likemode_login.js
 * 
 * @author:     Patryk Rzucidlo [@ptkdev] <info@ptkdev.it> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 * @version:    0.1
 * @changelog:  0.1 initial release
 *
 */
async function start_login() {
    let login = require(__dirname + '/modules/login.js')(bot, config, utils);
    utils.logger("[INFO]", "login", "loading...");
    login.open_loginpage(utils);
    utils.sleep(utils.random_interval(4, 8));
    login.set_username();
    utils.sleep(utils.random_interval(4, 8));
    login.set_password();
    utils.sleep(utils.random_interval(4, 8));
    login.submit();
    utils.sleep(utils.random_interval(4, 8));
    login_status = await login.submitverify();
    utils.logger("[INFO]", "login", "login_status is " + login_status);
    bot = login.get_bot();
}


/**
 * Start Bot (flow) 
 * =====================
 *
 * @author:     Patryk Rzucidlo [@ptkdev] <info@ptkdev.it> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 * @version:    0.1
 * @changelog:  0.1 initial release
 *
 */
start_login();
utils.sleep(utils.random_interval(4, 8));
if (login_status == 1) {
    start_twofa_check();
    utils.sleep(utils.random_interval(4, 8));
    if (pin_status == 1) {
        utils.sleep(utils.random_interval(4, 8));
        start_twofa();
        if (twofa_status == 1) {
            switch_mode(bot, config, utils);
        }
    } else {
        switch_mode(bot, config, utils);
    }
}